File: Wheel.java
import java.util.ArrayList;
import java.util.Random;

//
// HEADER HERE
//
//
public class Wheel 
{

   //
   // lToR and rToL arraylists.
   //
   private final ArrayList<Integer> lToR;
   private final ArrayList<Integer> rToL;
   
   private final int                nRec;	   
   private final int            notchLoc;
   private       int              curPos;
   private final long               seed;
   
   private final static boolean 
                            DEBUG = true;

   private void loadLToR()
   {
      final int rRange   = nRec - 1;
      Random r = new Random(rRange);
      r.setSeed(seed);
      
      for (int i = 0; i<nRec; ++i) 
                             lToR.add(i);
      for (int i = 0; i<nRec; ++i)
      {
         final int iSav = lToR.get   (i);
         int j = r.nextInt(rRange);
         
         
         //
         // Swap the entries at i
         // and j.
         //
         final int jSav = lToR.get(j);
         lToR.set(i,jSav);
         lToR.set(j,iSav);

      }
         
      if (DEBUG)
      {
         System.out.print  ("nRec = ");
         System.out.print       (nRec);
         System.out.print (" nLoc = ");
         System.out.print   (notchLoc);
         System.out.print (" sPos = ");
         System.out.print     (curPos);
         System.out.print("  seed = ");
         System.out.print       (seed);
         System.out.println      ("L");
         
         for (int i = 0; i<nRec; ++i)
         {
            final int j = lToR.get(i);
            System.out.print  ("i = ");
            System.out.print       (i);
            if(i<10) 
                 System.out.print(" ");
            System.out.print("  j = ");
            System.out.println     (j); 
            if (i==j)
            {
               System.out.println
                          ("MATCH!!!");
//               System.exit         (i);
            }
         }
         for (int i = 0; i<nRec; ++i)
         {
            boolean found = false;
            for (int j = 0; j<nRec && (found == false); ++j)
               found = ( j == lToR.get(i) );
             if (!found)
            {
               System.out.print(i);
               System.out.print(" NOT FOUND ");
               System.out.print("  ");
               System.out.println
                          ("ERROR!!!");
               System.exit         (i);
            }                  
         }            
      }
      
   }
   
   //
   // O(n squared)
   // function to load
   // rToL
   //
   private void loadRToL()
   {      
      for (int i = 0; i < nRec; ++i)
      {
         boolean found = false;
         int at = 0;
         while (!found)
         {
            final int j = lToR.get(at);
            if ( i == j )
            {
               found    =     true;
               rToL.add(at);
            }
            else ++at;
         }
      }
      if(DEBUG)
         for (int i = 0; i<nRec; ++i)
         {
            boolean found = false;
            for (int j = 0; j<nRec && (found == false); ++j)
               found = ( j == rToL.get(i) );
             if (!found)
            {
               System.out.print(i);
               System.out.print(" NOT FOUND ");
               System.out.print("  ");
               System.out.println
                          ("ERROR!!!");
               System.exit         (i);
            }                  
         } 
   }

   //
   // Converts wheel entry 
   // to a value to be added
   // to or subtracted from
   // the character position
   // mod nRec.
   //
   private void convertToOffset()
   {
      for (int i =0; i<nRec; ++i)
      {
         int offset = lToR.get(i) - i;
         lToR.set(i,offset);
         offset = rToL.get(i) - i;
         rToL.set(i,offset);
      }
   }

   //
   // nEnts = size of Wheel
   // seedIn = randome number seed 
   //   to produce unique connections
   // cPos = current wheel position
   // nPos = notch position
   //
   public Wheel(int nEnts, 
                long seedIn, 
                int cPos, 
                int nPos)
   {
      nRec =             nEnts;
      notchLoc =          nPos;
      curPos   =          cPos;
      seed     =        seedIn;
      lToR = new ArrayList<>();
      rToL = new ArrayList<>();
      loadLToR              ();
      loadRToL              ();
      convertToOffset       ();
   }
	
   //
   // Returns true if at 
   // notch, false otherwise.
   //
   public boolean atNotch()
   {
      boolean aN = ( curPos == notchLoc );
      return aN;
   }
   
   public void resetCur( final int cP)
   {
      curPos = cP;
   }
   
   public int getLtoR(final int i)
   {
      final int index = (curPos + i) % nRec;
      int val = lToR.get(index);
      val += i;
      if ( val < 0 ) val += nRec;
      else if (val >= nRec) val-= nRec;
      return val;
   }
   
   public int getRtoL(final int i)
   {
      final int index = (curPos + i) % nRec;
      int val = rToL.get(index);
      val += i;
      if ( val < 0 ) val += nRec;
      else if (val >= nRec) val-= nRec;
      return val;
   }
   
   public void advance() 
   {
      ++curPos;
      curPos = (curPos >= nRec) ? 0 : curPos;
   }
   
   public void reverse() 
   {
      --curPos;
      curPos = (curPos < 0) ? nRec-1 : curPos;
   }
}

File: wheel.cpp
#include "wheel.h"
#include<cstdlib>


using namespace std;

//wheel constructor initializer for number of entries, seed, and current position
Wheel::Wheel(
    const short       nEnts,
    const unsigned int seed,
    const short        cPos) : nRec(nEnts), curPos(cPos)
{

    srand(seed);

    notch = rand() % nRec;

    loadLToR();
    loadRToL();
    convertToOffset();

}

// initializes the ltoR vector with random values from 0 to 9
void Wheel::loadLToR() {

    for (short i = 0; i < nRec; ++i) {
        lToR.push_back(i);
    }
    for (short i = 0; i < nRec; ++i) {

        short j = rand() % nRec;
        swap(lToR[i], lToR[j]);
    }

}

// initializes the rtoL vector with random values from 0 to 9
void Wheel::loadRToL() {

    for (short i = 0; i < nRec; ++i) {
        rToL.push_back(0);
    }
    for (short i = 0; i < nRec; ++i) {
        rToL[lToR[i]] = i;
    }
}


// gets absolute values in ltoR and rToL to offsett realtive to their indices
void Wheel::convertToOffset() {
    for (short i = 0; i < nRec; ++i) {
        short offsett = lToR.at(i) - i;
        lToR[i] = offsett;
        offsett = rToL.at(i) - i;
        rToL[i] = offsett;
    }
}

//computes the transformed value from ltoR using the current position and index
short Wheel::getLtoR(short i) {
    const short index = (curPos + i) % nRec;
    short val = lToR.at(index);
    val += i;
    if (val < 0) val += nRec;
    else if (val >= nRec) val -= nRec;
    return val;

}

//computes the transformed value from rtoL using the current position and index
short Wheel::getRtoL(short i) {
    const short index = (curPos + i) % nRec;
    short val = rToL.at(index);
    val += i;
    if (val < 0) val += nRec;
    else if (val >= nRec) val -= nRec;
    return val;

}

// clear the vector
Wheel::~Wheel() {
    lToR.clear();
    rToL.clear();
}